modules/context/repo.go

```diff
@@ -602,6 +609,9 @@ func RepoAssignment(ctx *Context) (cancel context.CancelFunc) {
 		ctx.ServerError("RepoAssignment Invalid repo "+repo_model.RepoPath(userName, repoName), err)
 		return
 	}
+	if ctx.Repo.GitRepo != nil {
+		ctx.Repo.GitRepo.Close()
+	}
```

Comment:
I didn't see where the `ctx.Repo.GitRepo` was set.

Do you mean `gitRepo` above?

![image](https://user-images.githubusercontent.com/2114189/164355346-f5621156-43fb-4eb0-9543-e8441b2fe1ad.png)

Reply:
The usage of `ctx.Repo.GitRepo` should be refactored together. For example, the `GetGitRefs` may leave a closed & dangling `ctx.Repo.GitRepo`.

There should be a rule about how to use `ctx.Repo.GitRepo`:
* when it is closed, it must be set to nil.
* when it should be nil but now, there should be an error log to remind developers that there is a bug (forgot to close the GitRepo)


![image](https://user-images.githubusercontent.com/2114189/164356186-36c5eb79-2822-419a-99d5-5a8460108cce.png)

Reply:
that's right

`defer ctx. ....GitRepo.Close() ` != gitRepo.Close()

Reply:
I can not understand what "right" means.

For the screenshot, do you think it is right? After `defer ctx...GetRepo.Close()`, shouldn't the code set `ctx...GitRepo` to nil? Otherwise some other functions may use closed GitRepo again.

Reply:
the scope/layer that open a repo should ensure it gets closed - so here it's fine :)